### ISSUE TITLE
./inject.pl is running code from installed tree and not own source

### ISSUE THREAD

**Issue by nieder** [fink/fink#266]
This has come up several times in the mailing lists, where the Services.pm , where the code to deal with macOS vs darwin versioning schemes is different

https://github.com/fink/fink/blob/867a27fedc2c0160bfe14ad139dc455effd0ca3e/perlmod/Fink/Services.pm#L1442-L1454

During updates, the ./inject.pl runs crash because existing installs don't know about a new mismatch Apple introduced, and therefore they exit.

I installed a clean tree to /opt/sw-inject, made sure my terminal only knew about this new tree, and added a `die "debug text"` debug line inside Services.pm::get_osx_vers() for both `/opt/sw-inject/lib/perl5/Fink/Services.pm` and the `Services.pm` in my git pull. When I run `./inject.pl`, I see the following:

```
Modifying package description...
/bin/chmod 644 /opt/sw-inject/fink/dists/local/injected/finkinfo/*.*
Installing package...

/opt/sw-inject/bin/fink install fink-0.45.99.git-20240327.0241
Checking Services.pm in the installed tree. at /opt/sw-inject/lib/perl5/Fink/Services.pm line 1429.
### execution of /opt/sw-inject/bin/fink failed, exit code 29
```
Both the text and the referenced line number point to ./inject.pl using the code from the existing installed tree, not from the git pull.

@TheSin-

**Comment by TheSin-** [comment]
os I think the issue isn't with inject it's self.  I think it called fink (even the new version in the inject dir) but since it's a new call the `use lib` won't be set anymore.

I'm not sure how to overcome this without some way for fink to know it's not installed yet and use a different lib dir.

**Comment by nieder** [comment]
Adding comment to say that others have been getting the actual inject failure due to the $sw_vers and $darwin_osx mismatch on macOS 11 and 13 (where mismatches happen). However, my debugging was on macOS 10.14.6 with system-perl-5.18.4, so it's not something due to newer perl handling imports differently.

**Comment by nieder** [comment]
The failure is happening when %p/bin/fink run from inside Bootstrap.pm::inject_package() tries to install the newly created tarball from the inject tree:

https://github.com/fink/fink/blob/867a27fedc2c0160bfe14ad139dc455effd0ca3e/perlmod/Fink/Bootstrap.pm#L395

but $inject_tree/fink doesn't exist (it's just fink.in).

**Comment by cooljeanius** [comment]
> The failure is happening when %p/bin/fink run from inside Bootstrap.pm::inject_package() tries to install the newly created tarball from the inject tree:
> 
> https://github.com/fink/fink/blob/867a27fedc2c0160bfe14ad139dc455effd0ca3e/perlmod/Fink/Bootstrap.pm#L395
> 
> but $inject_tree/fink doesn't exist (it's just fink.in).

`git blame` says that Dave Vasilevsky was the last to touch that code in 2f41b2b... that's @vasi, right?

**Comment by nieder** [comment]
`Fink::Engine::&cmd_install` seems to be the internal function eventually being used. Is there something like 
``` perl
perl -I$INJECT_TREE -MFink::Engine ' "&cmd_install $package-$packageversion-$packagerevision"'
```
?

Though that's still trying to execute as an external program. Bootstrap.pm already has `use Fink::Engine`. There's code elsewhere in Bootstrap.pm for `Fink::Engine::cmd_install(@elist, @addlist);`. Can we use something like this instead of the problematic `&execute("$bpath/bin/fink install ...")` command?

**Comment by nieder** [comment]
Progress if I run `Fink::Package->require_packages();` before `Fink::Engine::cmd_install(...)`, but still a failure:
```
Setting runtime build-lock...
dpkg-deb -b /opt/sw-0.45.6-for-inject-debug/src/fink.build/root-fink-buildlock-fink-0.45.99.git-20250723.1139 /opt/sw-0.45.6-for-inject-debug/src/fink.build
dpkg-deb: building package `fink-buildlock-fink-0.45.99.git-20250723.1139' in `/opt/sw-0.45.6-for-inject-debug/src/fink.build/fink-buildlock-fink-0.45.99.git-20250723.1139_2025.07.23-06.39.12_darwin-@ARCHITECTURE@.deb'.
Installing build-lock package...
/opt/sw-0.45.6-for-inject-debug/bin/dpkg-lockwait -i /opt/sw-0.45.6-for-inject-debug/src/fink.build/fink-buildlock-fink-0.45.99.git-20250723.1139_2025.07.23-06.39.12_darwin-@ARCHITECTURE@.deb
/sw/bin/dpkg: error processing /opt/sw-0.45.6-for-inject-debug/src/fink.build/fink-buildlock-fink-0.45.99.git-20250723.1139_2025.07.23-06.39.12_darwin-@ARCHITECTURE@.deb (--install):
 package architecture (darwin-@ARCHITECTURE@) does not match system (darwin-x86_64)
Errors were encountered while processing:
 /opt/sw-0.45.6-for-inject-debug/src/fink.build/fink-buildlock-fink-0.45.99.git-20250723.1139_2025.07.23-06.39.12_darwin-@ARCHITECTURE@.deb
### execution of /opt/sw-0.45.6-for-inject-debug/bin/dpkg-lockwait failed, exit code 1
Can't set build lock for fink (0.45.99.git-20250723.1139)
```
@ARCHITECTURE@ was not resolved.

For comparison, here's what happens when 0.45.3 is updated to 0.45.6 via inject.pl, which uses the problematic `%p/bin/fink install fink...` command:
```
The following package will be installed or updated:
 fink
Setting runtime build-lock...
dpkg-deb -b /opt/sw-0.45.3-for-0.45.6-inject/src/fink.build/root-fink-buildlock-fink-0.45.6-141 /opt/sw-0.45.3-for-0.45.6-inject/src/fink.build
dpkg-deb: building package `fink-buildlock-fink-0.45.6-141' in `/opt/sw-0.45.3-for-0.45.6-inject/src/fink.build/fink-buildlock-fink-0.45.6-141_2025.07.23-22.12.30_darwin-x86_64.deb'.
Installing build-lock package...
/opt/sw-0.45.3-for-0.45.6-inject/bin/dpkg-lockwait -i /opt/sw-0.45.3-for-0.45.6-inject/src/fink.build/fink-buildlock-fink-0.45.6-141_2025.07.23-22.12.30_darwin-x86_64.deb
```

**Comment by nieder** [comment]
So, once the files from the new deb have been installed, how do we get the new files/pkg noticed as installed? Do we need to run `PostInstScript` or something similar? If so, how can this be done within perl without calling the `fink` command itself that got us into this mess? Does this involve `postinstall.pl::pre_scanpackages()` as seen in fink.info::PostInstScript?

Also, to remind myself: all the real changes above happen in ./inject.pl and Bootstrap.pm, so in _theory_ it shouldn't affect installing any packages the normal way.

**Comment by kristibektashi** [comment]
Any progress in this?

**Comment by kristibektashi** [comment]
Is there any workaround? I'm running latest Tahoe and I'm getting the error `arch: /usr/bin/perl5.30 isn't executable` when running fink, which I believe is supposed to be fixed. ./inject.pl also fails with error `### execution of /usr/bin/perl failed, exit code 29`, and I can only assume the above is the reason why.

**Comment by nieder** [comment]
For fixing the ./inject.pl problem that it runs an installed Fink rather than the tarball Fink, I've debugged as much as my limited perl knowledge lets me.
Near as I can tell, the changes/edits I listed above eventually fail with the new packages installed through the correct paths, but the database not updated to know they've been installed. Someone that actually knows Perl needs to pick up it up from there, or find the correct method if what I found is totally fubar.

**Comment by nieder** [comment]
> Is there any workaround? I'm running latest Tahoe and I'm getting the error `arch: /usr/bin/perl5.30 isn't executable` when running fink, which I believe is supposed to be fixed. ./inject.pl also fails with error `### execution of /usr/bin/perl failed, exit code 29`, and I can only assume the above is the reason why.

What perl version does your Tahoe have? Is the OS version 26.4.1 and darwin-25.4.0?

**Comment by kristibektashi** [comment]
> > Is there any workaround? I'm running latest Tahoe and I'm getting the error `arch: /usr/bin/perl5.30 isn't executable` when running fink, which I believe is supposed to be fixed. ./inject.pl also fails with error `### execution of /usr/bin/perl failed, exit code 29`, and I can only assume the above is the reason why.
> 
> What perl version does your Tahoe have? Is the OS version 26.4.1 and darwin-25.4.0?

Yep that's the exact one, it has Perl 5.34 (`perl`, `perl5` and `perl5.34` all exist and point to 5.34, `perl5.30` doesn't exist at all)
